python/ray/data/dataset.py

```diff
@@ -827,14 +828,80 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
+            >>> ray.data.range(100).agg("sum")
+            >>> ray.data.range(100).agg(["sum", "mean"])
+            >>> ray.data.from_items([
+                        {"A": x, "B": x**2} for x in range(100)]) \
+                    .agg({
+                        "A": ["mean", "std"],
+                        "B": ["min", "max"],
+                    })
+
+        Args:
+            aggs: Names of aggregations to do. This can be:
+                - str: The name of a single aggregation that we want to apply
+                    to all columns.
+                - List[str]: A list of the names of aggregations that we want
+                    to apply to all columns.
+                - Dict[str, str]: A map from column names to the names of the
+                    aggregations that we wish to apply to those specific
+                    columns. This can only be applied to Arrow Datasets.
+
+        Returns:
+            If the input dataset is a simple dataset then the output is
+            a tuple of (agg1, agg2, ...) where each tuple element is
+            the corresponding aggregation result.
+            If the input dataset is an Arrow dataset then the output is
+            an ArrowRow where each column is the corresponding
+            aggregation result.
+            When `aggs` is given as a list, the aggregation columns are ordered
+            in accordance with the aggregation list ordering (aggs, cols):
+                [agg1, agg2]
+            on a dataset with columns col1 and col2 would yield aggregation
+            columns:
+                col1_agg1, col2_agg1, col1_agg2, col2_agg2
+            When `aggs` is given as a dict, the aggregation columns are ordered
+            in accordance with the depth-first dict ordering (cols, aggs):
+                {col1: [agg1, agg2], col2: [agg2, agg1]}
+            would yield aggregation columns:
+                col1_agg1, col1_agg2, col2_agg2, col2_agg1
+            If the dataset is empty, return None.
+        """
+        aggs_ = []
+        if isinstance(aggs, str):
+            aggs = [aggs]
+        if isinstance(aggs, list):
+            aggs = ((None, aggs), )
+        elif isinstance(aggs, dict):
+            aggs = aggs.items()
+        else:
+            raise TypeError(
+                "`aggs` must be a string, list of strings, or a dict, "
+                f"got {aggs}.")
+        for on, on_aggs in aggs:
+            for agg in on_aggs:
+                on = self._check_and_normalize_agg_on(on)
+                if not isinstance(on, list):
+                    on = [on]
+                try:
+                    aggs_.extend([AGGS_TABLE[agg](on_) for on_ in on])
+                except KeyError as e:
+                    raise ValueError(f"{e!s} is not a supported aggregation: "
+                                     f"{list(AGGS_TABLE.keys())}")
+        return self.apply_aggregations(*aggs_)
+
+    def apply_aggregations(self, *aggs: AggregateFn) -> U:
```

Comment:
should we unify `agg` and `apply_aggregations` by accepting both AggregateFn and strs instead of having two functions, like https://pandas.pydata.org/docs/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html?

Reply:
There's a bit of an impedance mismatch here for the dict argument in `agg`. An instance of `AggregateFn` already has its column selection hardcoded, while the mapping from column name to aggregations that we want to apply is basically mapping `column: str` --> `agg_cls: type[AggregateFn]`, and then we use that mapping to create concrete aggregation instances: `agg_cls(on=column)`. I.e., giving `{"col1": [Mean()]}` doesn't make sense for the current `AggregateFn` API, since the `on="col1"` selection needs to be given to `Mean` on construction, not after.

I can explore a different API/model for the aggregation functions so the column selection can be specified at call time, that could have some other advantages too!

Reply:
Yea, `{"col1": [Mean()]}` doesn't make sense and I don't think we need to support here. I think what I'm trying to say is should we just have `def agg(self, aggs: Union[str, List[str], Dict[str, str], AggregateFn, List[AggregateFn]]):`

Reply:
That sounds good! A user may want that dict of col --> agg funcs API in the future, but we can hold off on it for now.

Reply:
In order to support `AggregateFn` and `List[AggregateFn]`, I needed to make a null `on` overridable, which then made it trivial to support `{"col1": [Mean()]}`.

Reply:
`In order to support AggregateFn and List[AggregateFn], I needed to make a null on overridable,` why is that?

Also I'm not sure if we need to support both `{"col1": ["mean"]}` and `{"col1": [Mean()]}`. Maybe we should support `{"col1": [Mean, Max, AggregateFnClass]}`? Also through this way, people cannot specify std with a different ddof so in some cases they still need to use the old api.

Reply:
>why is that?

In order to support `AggregateFn` instances, you need to be able to override the null `on` attribute with the normalized `on` attribute, e.g. `None` --> list of all column names. Ideally we would move the `on` normalization to `AggregateFn`, but the normalization requires context from the `Dataset` so that's not quite doable.

>Also I'm not sure if we need to support both {"col1": ["mean"]} and {"col1": [Mean()]}. Maybe we should support {"col1": [Mean, Max, AggregateFnClass]}? Also through this way, people cannot specify std with a different ddof so in some cases they still need to use the old api.

I think that we should still have the string shorthand for the sake of convenience, and for those that are used to the Pandas aggregation API. It's particularly convenient if you're just looking for the default configuration of an aggregation, which should be the common case. Us accepting instances of `AggregateFn` then allows for that custom aggregation configuration, e.g. with `ddof=0` for `std`.